### PR TITLE
SDI-493 Alert if size of index does not match the number of prisone in NOMIS

### PR DESCRIPTION
**Pull request by steverendell**
Create comparison endpoint

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt [hunk 1]
```diff
@@ -61,6 +66,103 @@ class GlobalSearchService(
     }
   }
 
+  data class NomisContext(var nomisFinished: Boolean, var num: String?)
+
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt:70]: This gets around the 10000 limit by using the 'scroll' api which should be able to get the whole index 10000 at a time

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt [hunk 2]
```diff
@@ -61,6 +66,103 @@ class GlobalSearchService(
...
+  fun compareIndex(): Pair<List<String>, List<String>> {
+    val onlyInIndex = mutableListOf<String>()
+    val onlyInNomis = mutableListOf<String>()
+
+    val allNomis =
+      prisonerIndexService.getAllNomisOffenders(0, Int.MAX_VALUE).offenderIds!!.sortedBy { it.offenderNumber }
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt:76]: This is pretty hefty, I thought Id see if it breaks dev before getting sophisticated about streaming etc.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt [hunk 3]
```diff
@@ -61,6 +66,103 @@ class GlobalSearchService(
...
+    iter: Iterator<OffenderId>
+  ) {
+    allIndex.forEach { index: String ->
+      if (nomisContext.nomisFinished) {
+        onlyInIndex.add(index)
+      } else {
+        while (index > nomisContext.num!!) {
+          onlyInNomis.add(nomisContext.num!!)
+          if (iter.hasNext()) {
+            nomisContext.num = iter.next().offenderNumber
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt:134]: Ill improve this var name 'num'

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt [hunk 4]
```diff
@@ -212,10 +212,7 @@ class PrisonerSearchService(
   }
 
   private fun fieldMatch(field: String, value: String): BoolQueryBuilder {
-    return QueryBuilders.boolQuery()
-      .must(
-        "$field", value
-      )
+    return QueryBuilders.boolQuery().must(field, value)
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/PrisonerSearchService.kt:215]: intellij warnings

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/GlobalSearchResource.kt [hunk 5]
```diff
@@ -86,4 +90,23 @@ class GlobalSearchResource(
...
+  fun compareIndex() {
+    val start = System.currentTimeMillis()
+    val (onlyInIndex, onlyInNomis) = globalSearchService.compareIndex()
+    val end = System.currentTimeMillis()
+    telemetryClient.trackEvent(
+      "index-report",
+      mapOf(
+        "onlyInIndex" to onlyInIndex.toString(),
+        "onlyInNomis" to onlyInNomis.toString(),
+        "timeMs" to (end - start).toString(),
```
Comment (steverendell) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/resource/GlobalSearchResource.kt:105]: the onlyInIndex or onlyInNomis could be huge - ill truncate them if so

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt [hunk 6]
```diff
@@ -61,6 +66,103 @@ class GlobalSearchService(
...
+    searchRequest.scroll(scroll)
+    return searchClient.search(searchRequest)
+  }
+
+  private fun processIndexScroll(
+    allIndex: List<String>,
+    nomisContext: NomisContext,
+    onlyInIndex: MutableList<String>,
+    onlyInNomis: MutableList<String>,
+    iter: Iterator<OffenderId>
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/GlobalSearchService.kt:125]: This all seems a bit complicated, manually controlling iterators. Can't we just use [minus](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/minus.html)?

**Review by mikehalmamoj — APPROVED** [review]

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**